firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -159,66 +163,101 @@ public static FirebaseInstallations getInstance(@NonNull FirebaseApp app) {
-  @VisibleForTesting
-  String getApplicationId() {
-    return firebaseApp.getOptions().getApplicationId();
-  }
-
-  /** Returns the nick name of the {@link FirebaseApp} of this {@link FirebaseInstallations} */
-  @VisibleForTesting
-  String getName() {
-    return firebaseApp.getName();
-  }
-
-  /**
-   * Returns the {@link PersistedFidEntry} from shared prefs.
-   *
-   * @throws {@link FirebaseInstallationsException} when shared pref is empty or {@link
-   *     PersistedFidEntry} is in error state.
-   */
-  private PersistedFidEntry getPersistedFid() throws FirebaseInstallationsException {
-    PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
-    if (persistedFidMissingOrInErrorState(persistedFidEntry)) {
-      throw new FirebaseInstallationsException(
-          "Failed to get existing fid.", FirebaseInstallationsException.Status.CLIENT_ERROR);
+  private Task<String> addGetIdListener() {
+    TaskCompletionSource<String> taskCompletionSource = new TaskCompletionSource<>();
+    StateListener l = new GetIdListener(taskCompletionSource);
+    synchronized (lock) {
+      listeners.add(l);
     }
-    return persistedFidEntry;
+    return taskCompletionSource.getTask();
   }
 
-  private static boolean persistedFidMissingOrInErrorState(PersistedFidEntry persistedFidEntry) {
-    return persistedFidEntry == null
-        || persistedFidEntry.getRegistrationStatus() == RegistrationStatus.REGISTER_ERROR;
+  private Task<InstallationTokenResult> addGetAuthTokenListener(
+      @AuthTokenOption int authTokenOption) {
+    TaskCompletionSource<InstallationTokenResult> taskCompletionSource =
+        new TaskCompletionSource<>();
+    StateListener l = new GetAuthTokenListener(utils, taskCompletionSource);
+    synchronized (lock) {
+      if (authTokenOption == FORCE_REFRESH) {
+        shouldRefreshAuthToken = true;
+      }
+      listeners.add(l);
+    }
+    return taskCompletionSource.getTask();
   }
 
-  @NonNull
-  private static <F, T> Continuation<F, T> orElse(@NonNull Supplier<T> supplier) {
-    return t -> {
-      if (t.isSuccessful()) {
-        return (T) t.getResult();
+  private void triggerOnStateReached(PersistedFidEntry persistedFidEntry) {
+    synchronized (lock) {
+      Iterator<StateListener> it = listeners.iterator();
+      while (it.hasNext()) {
+        StateListener l = it.next();
+        boolean doneListening = l.onStateReached(persistedFidEntry);
+        if (doneListening) {
+          it.remove();
+        }
       }
-      return supplier.get();
-    };
+    }
   }
 
-  @NonNull
-  private static <F, T> Continuation<F, T> awaitFidRegistration(
-      @NonNull Supplier<T> supplier, AwaitListener listener) {
-    return t -> {
-      // Waiting for Task that registers FID on the FIS Servers
-      listener.await(AWAIT_TIMEOUT_IN_SECS, TimeUnit.SECONDS);
-      return supplier.get();
+  private final Runnable doRegistration() {
```

Comment:
I wouldn't bother wrapping this in a method like this. Either inline it at the field declaration site (`Runnable doRegistration = () -> { /* code here */ }` or make `doRegistration` the actual runnable and just chuck a method reference on the executor (`execute(this::doRegistration)`).

Reply:
Done. PTAL.